contracts/contracts/BLSWallet.sol

```diff
@@ -14,31 +14,75 @@ import "./interfaces/IWallet.sol";
 contract BLSWallet is Initializable, IBLSWallet
 {
     uint256 public nonce;
+    bytes32 public recoveryHash;
+    bytes32 pendingRecoveryHash;
+    uint256 pendingRecoveryHashTime;
+    bytes public approvedProxyAdminFunction;
+    bytes pendingPAFunction;
+    uint256 pendingPAFunctionTime;
 
     // BLS variables
     uint256[4] public blsPublicKey;
+    uint256[4] pendingBLSPublicKey;
+    uint256 pendingBLSPublicKeyTime;
     address public trustedBLSGateway;
+    address pendingBLSGateway;
+    uint256 pendingGatewayTime;
+
+    event PendingRecoveryHashSet(
+        bytes32 pendingRecoveryHash
+    );
+    event PendingBLSKeySet(
+        uint256[4] pendingBLSKey
+    );
+    event PendingGatewaySet(
+        address pendingGateway
+    );
+    event PendingProxyAdminFunctionSet(
+        bytes pendingProxyAdminFunction
+    );
+
+    event RecoveryHashUpdated(
+        bytes32 oldHash,
+        bytes32 newHash
+    );
+    event BLSKeySet(
+        uint256[4] oldBLSKey,
+        uint256[4] newBLSKey
+    );
+    event GatewayUpdated(
+        address oldGateway,
+        address newGateway
+    );
+    event ProxyAdminFunctionApproved(
+        bytes approvedProxyAdmin
+    );
```

Comment:
Something to consider for the future would be to break this down into separate abstract component contracts that manage the state/pending/recovery of these sensitive fields. i.e. `blsPublicKey` and its pending fields, events, mutators, and set pending calls, then have `BLSWallet` inherit them. This would reduce the code size/complexity of `BLSWallet.sol`, and make it easier to break out isolated unit test suites for them.

Reply:
Agreed, at that stage it would be good to factor in upgradability of wallets (bls and potentially non-bls). Which needs to consider the inheritance order since memory locations when upgrading are important. -> Created #73